### ISSUE TITLE
Pack individual functions using the new packageFunction hook

### ISSUE THREAD

**Issue by m3rls** [elastic-coders/serverless-webpack#60]
Hi,

I've noticed that when trying to deploy individual functions (with `package -> individually: true` in serverless.yml) the change made in https://github.com/elastic-coders/serverless-webpack/pull/48 has fixed the exception (thank you!), but the function isn't being packed.  

In my use case, this leads to a ~19MB .zip being uploaded compared to a ~600kb .zip when I do a full deploy.

[A Pull Request](https://github.com/serverless/serverless/pull/2561) was merged and released along with the serverless 1.2.0 milestone yesterday that added a `before:deploy:function:packageFunction` hook, which I have (naively and probably incorrectly) added to `index.js` as follows:

```js
      'before:deploy:function:packageFunction': () => BbPromise.bind(this)
        .then(this.validate)
        .then(this.compile),
```

So now when I run `serverless deploy function --function <myFunction>` it packages it (as illustrated below):

### Without hook:
```
➜  api git:(master) ✗ serverless deploy function --function myFunction
Serverless: Deploying function: myFunction...
Serverless: Packaging function: myFunction...
Serverless: Uploading function: myFunction (19.9 MB)...
Serverless: Successfully deployed function: myFunction
```

### With hook:
```
➜  api git:(master) ✗ serverless deploy function --function myFunction
Serverless: Deploying function: myFunction...
Serverless: Bundling with Webpack...
Time: 2337ms
     Asset     Size  Chunks             Chunk Names
handler.js  1.92 MB       0  [emitted]  main
Serverless: Packaging function: myFunction...
Serverless: Uploading function: myFunction (606.73 KB)...
Serverless: Successfully deployed function: myFunction
```

Would it be possible to please have support for this hook added at some point?  I'm happy to make a pull request for it if the way I've added it is acceptable, but didn't want to just assume it was correct.

Thanks

**Comment by activefx** [comment]
@m3rls would you be willing to share your webpack config? I'm having trouble packaging functions individually and would appreciate it

**Comment by m3rls** [comment]
@activefx sure, it's pretty minimal:

```js
module.exports = {
  entry: './handler.js',
  target: 'node',
  module: {
    loaders: [
      { test: /\.json$/, loader: 'json' }, {
        test: /\.js$/,
        loaders: ['babel'],
        include: __dirname,
        exclude: /node_modules/,
      }
    ]
  }
};
```

I've had to use the latest version of `serverless-webpack` from this repo rather than the one from npmjs.com until a new release is pushed up, due to https://github.com/elastic-coders/serverless-webpack/pull/48

**Comment by activefx** [comment]
Thank you for sharing @m3rls, I must be missing something else then

**Comment by thenikso** [comment]
it would be great indeed. I've checked out the new "packageFunction" related methods in Serverless and adding a hook on `deploy:function:packageFunction` should do the trick. 

Anyone willing to give it a try?

**Comment by johnf** [comment]
The solution @m3rls described works for me. I've created a pull request in #91

**Comment by HyperBrain** [comment]
This has to be checked again as soon as #130 is merged, to see if there is still anything to do here.

**Comment by HyperBrain** [comment]
PR #130 has been merged and published (current version is 2.2.0).
There should be full support of `serverless deploy function --function=<myfunction>` now. @m3rls , @johnf , @activefx - can you try if that works for you?

**Comment by johnf** [comment]
So all my files actually live in a src directory. My webpack config was dealing with this previously.
If I move everything to the root it works prefectly.

#176 is a possible fix. If others agree I'll update docs etc.

in serverless.yaml I added
``` yaml
custom:
  webpackBasePath: './src'
```

**Comment by HyperBrain** [comment]
@johnf I think you did not declare your handlers properly in your serverless.yml. Serverless itself relies on the fact that you specify your handlers there relative to your service root directory.

In your case this would be:
```
function:
  myfunction:
    handler: src/handler.myFunc
```

The new package/deploy function support works with and without the webpack plugin. Imo, having a separate base path definition only for the webpack plugin would introduce a different behavior when the plugin is enabled or not.

Can you check if your handler definitions in your service point to the right handlers?

**Comment by johnf** [comment]
@HyperBrain thanks don't know why I didn't think if that it's fairly obvious. Have updated and looks good now

**Comment by HyperBrain** [comment]
BTW: I implemented the support of the service wide `package: individually: true`. It is available in the `v3.0.0-individual-packaging` branch. Can you test, if the branch works, if you set the flag at service level?
Please comment the results in #177 .

**Comment by HyperBrain** [comment]
I'll close this task as the problem is solved and individual packaging is now fully supported in `3.0.0-rc.2`
